[O365] | Migrate to CEL input

kcreddy opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

## What does this PR do?
Adds CEL input for fetching events from O365 Management API. Deprecates the existing o365audit input.

<!-- Mandatory
Explain here the changes you made on the PR.
-->

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally
- Run `elastic-package stack down && elastic-package build && elastic-package stack up -d -v && eval "$(elastic-package stack shellinit)" && elastic-package test system --generate  -v`
Produces:
```
.......
2023/07/12 13:01:10 DEBUG found 0 hits in logs-o365.audit-ep data stream: index_not_found_exception: no such index [logs-o365.audit-ep] Status=404
2023/07/12 13:01:11 DEBUG found 0 hits in logs-o365.audit-ep data stream: index_not_found_exception: no such index [logs-o365.audit-ep] Status=404
2023/07/12 13:01:12 DEBUG found 0 hits in logs-o365.audit-ep data stream
2023/07/12 13:01:13 DEBUG found 27 hits in logs-o365.audit-ep data stream
........
........
2023/07/12 13:02:10 DEBUG found 0 hits in logs-o365.audit-ep data stream
2023/07/12 13:02:11 DEBUG found 0 hits in logs-o365.audit-ep data stream
2023/07/12 13:02:12 DEBUG found 0 hits in logs-o365.audit-ep data stream
2023/07/12 13:02:13 DEBUG found 0 hits in logs-o365.audit-ep data stream
2023/07/12 13:02:14 DEBUG found 6 hits in logs-o365.audit-ep data stream
.........
.........
--- Test results for package: o365 - START ---
╭─────────┬─────────────┬───────────┬───────────┬────────┬───────────────╮
│ PACKAGE │ DATA STREAM │ TEST TYPE │ TEST NAME │ RESULT │  TIME ELAPSED │
├─────────┼─────────────┼───────────┼───────────┼────────┼───────────────┤
│ o365    │ audit       │ system    │ o365audit │ PASS   │ 27.106824625s │
│ o365    │ audit       │ system    │ cel       │ PASS   │ 26.534162792s │
╰─────────┴─────────────┴───────────┴───────────┴────────┴───────────────╯
--- Test results for package: o365 - END   ---
Done
```

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Supersedes #123
-->
- Closes https://github.com/elastic/integrations/issues/5045

## Screenshots

<!-- Optional
Add here screenshots presenting:
- Kibana UI forms presenting configuration options exposed by the integration
- dashboards with collected metrics or logs
-->

jamiehynds commented:
@kcreddy can you confirm the behaviour when a user of the current O365 integration updates to the new version that uses CEL? Will the switch to CEL be automatic without any user intervention or will httpjson be disabled and a user has re-enter their O365 API manually?

kcreddy commented:
> Will the switch to CEL be automatic without any user intervention or will httpjson be disabled and a user has re-enter their O365 API manually?

Users would have to setup the new CEL input manually. The input parameters are pretty similar to what they have been using with O365 Audit input ie., `Azure Client ID, Tenant ID, Secret`, etc,. The package and datastream the events end up is also same. i.e., `data_stream.dataset: o365.audit`. In other words we are not creating a new package or a datastream.

The screenshot of setup looks like this:
<img width="856" alt="Screenshot 2023-06-21 at 10 33 29 AM" src="https://github.com/elastic/integrations/assets/11301409/4ea4e3cc-19ed-4baf-9bcf-b1c342bcfd9e">
<img width="644" alt="Screenshot 2023-06-21 at 10 33 44 AM" src="https://github.com/elastic/integrations/assets/11301409/07c3d5d6-1747-47c1-bfc9-f572e4de2a54">
------------------------------------------------------------------------------


> can you confirm the behaviour when a user of the current O365 integration updates to the new version that uses CEL?

After getting suggestions from @andrewkroh, the behaviour would result in some duplicate events for users. That is because when the users upgrade the integration, the index rollover happens. So, even the existing `fingerprint` processor inside ingest pipeline on event's `Id` field would still create duplicates for last few days (up until 7 days back). We have the leverage to decide how far back we want to start pulling the events. We could also give the users ability to choose their start time. [Reference](https://learn.microsoft.com/en-us/office/office-365-management-api/office-365-management-activity-api-reference#working-with-the-office-365-management-activity-api:~:text=Both%20must%20be%20specified%20(or%20both%20omitted)%20and%20they%20must%20be%20no%20more%20than%2024%20hours%20apart%2C%20with%20the%20start%20time%20no%20more%20than%207%20days%20in%20the%20past.%20By%20default%2C%20if%20startTime%20and%20endTime%20are%20omitted%2C%20then%20the%20content%20available%20in%20the%20last%2024%20hours%20is%20returned.) 
The behaviour still needs to be tested though.

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-6621/detail/PR-6621/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-6621/detail/PR-6621/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-6621/detail/PR-6621/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-6621/detail/PR-6621/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_6621.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-06-21T05:09:03.866Z&rangeTo=2023-06-21T05:29:03.866Z&transactionName=BUILD+Ingest-manager%2Fintegrations%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=cb380a586804141b07749cf4820f170e&transactionId=2ef3eb454f027ac5)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-06-21T05:19:03.866+0000


* Duration: 15 min 37 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 24  |
| Skipped      | 0 |
| Total        | 24   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`1/1`) | :green_heart: 
Files | 100.0% (`1/1`) | :green_heart: 2.844
Classes | 100.0% (`1/1`) | :green_heart: 2.844
Methods | 100.0% (`16/16`) | :green_heart: 7.021
Lines | 80.994% (`733/905`) | :-1: -10.744
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

jamiehynds commented:
@kcreddy is the user prompted to input the credentials once they've upgraded the integration? We need to minimise the chances of a user upgrade the package, saving it and not realising that ingestion has stopped until they enter their Client ID & Secret. If we enable the datastream by default, I presume they won't be able to save the integration and will have to populate the details before continuing? We should also make this obvious in the changelog too. 

I'd suggest adjusting the description of the datastreams too:
Old datastream using O365 input: "Collect Office 365 Audit Logs - Deprecated"
New datastream using CEL: "Collect Office 365 Audit Logs via Graph API" (assuming we're using the Graph API?)

kcreddy commented:
Hey @jamiehynds,
Below is how the upgrade scenario is going to look like for the user:

https://github.com/elastic/integrations/assets/11301409/e188905b-4da5-4699-b3fd-96f34cbe74e1

Basically, the user will have to upgrade the integration policy too once the integration is upgraded to latest. Once that integration policy is updated, and since the new CEL input is enabled by default, the user will have to provide details for the CEL input to start. I have updated the title and description in the previous o365audit input to `Deprecated`. I also updated the `README` docs with the steps to follow when upgrading the integration from `< 1.7.0` to `>= 1.7.0`. Let me know if anything is missing or needs to be added/changed.

Also, during the tests, I noted that duplicates are not getting created. So, we are good there.

andrewkroh commented on packages/o365/manifest.yml:
```diff
@@ -29,7 +29,10 @@ policy_templates:
       - type: o365audit
-        title: "Collect Office 365 audit logs"
-        description: "Collect audit logs from Office 365 via the Management Activity API"
+        title: "Collect Office 365 audit logs - Deprecated. Please use the CEL input instead."
```
```suggestion
        title: "Collect Office 365 audit logs - Deprecated. Please disable this and use the CEL input instead."
```

I was thinking the message somewhere should tell them to disable it explicitly.

kcreddy replied:
Updated the manifest as suggested.

kcreddy commented:
/test

elasticmachine commented:
Pinging @elastic/security-external-integrations (Team:Security-External Integrations)

jamiehynds commented:
@kcreddy thanks for putting the video together - really helpful. 

When a user upgrades the the integration the 'upgrade integration policies' option is selected, but in this case the policy isn't updated. It's probably a bit misleading, but not sure we can solve for that. 

After upgrading the integration, a user has to go to integration policies to enable the new input and then upgrade the policy. I worry that users may not inattentively know to go to integration policies after the integration upgrade. 

I don't think people will know what CEL is. Could we adjust the wording of the new datastream to 'Collect Office 365 Audit Logs via Graph API' (or whatever API we're hitting)?

Am I correct in saying v1.17 of the integration will not include the deprecated input? So for new users using the O365 integration for the first time, will only have the option to use the CEL input, or will they also see the deprecated input too?

efd6 commented on packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,180 @@
+  base:
+    list_contents_url: "{{url}}/api/v1.0/{{azure_tenant_id}}/activity/feed/subscriptions/content?contentType="
+    list_contents_start_time: "{{initial_interval}}"
+
```
Add a
```
redact:
  fields: ~
```
stanza. We will log a warning if there is no redact config even if it is not required for the specific integration so that when it is required it's less likely to be missed.

Are there any fields in state that may be sensitive? (The auth secrets won't be in state, so you are fine there).

kcreddy replied:
Apart from the `client_secret` which is used for oauth, there are not other sensitive fields. I added the stanza 👍🏼

efd6 commented on packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,180 @@
+                                state.base.list_contents_url + content_type + "&PublisherIdentifier={{azure_tenant_id}}"  
+                                    + "&startTime=" + (now() - duration(state.base.list_contents_start_time)).format(time_layout.RFC3339) 
+                                    + "&endTime=" + ((now() - duration(state.base.list_contents_start_time) + duration("24h")).as(calc_end_time, 
+                                        calc_end_time <= now() ? calc_end_time : now())).format(time_layout.RFC3339)
```
```suggestion
                                        calc_end_time <= now ? calc_end_time : now)).format(time_layout.RFC3339)
```

Is there a reason to use `now()` over `now`? The difference between the two is that (static) `now` is the time the CEL evaluation started and `now()` is the time that the `now()` call was made. Generally you will want the static case.

kcreddy replied:
Got it. `now` makes sense. Updated the code to replace `now()` with `now`

kcreddy replied:
Hey @efd6 I reverted this change back to `now()` because `now` is being evaluated to the time when CEL evaluation initially started, remains the same after each poll. Not sure if its a bug or its meant to work that way.  

Example errors  when I used `now` (consecutive polls i.e., after `3 min`):
`Start time and end time must both be specified (or both omitted) and must be less than or equal to 24 hours apart, with the start time prior to  end time and start time no more than 7 days in the past. StartTime:2023-07-18T09:39:16Z, EndTime:2023-07-18T09:39:14Z`

`Start time and end time must both be specified (or both omitted) and must be less than or equal to 24 hours apart, with the start time prior to end time and start time no more than 7 days in the past. StartTime:2023-07-18T09:39:17Z, EndTime:2023-07-18T09:39:14Z`

Notice that the `endTime` which is evaluated to `now` remains the same.

efd6 replied:
Hmmmm. That's problematic. I will look into it. I'm surprised, and it's in language, and so non-trivial. I would suggest a work-around of adding this prefix to the program and using `state.now`.
```
state.with({"now": now()}).
```

efd6 replied:
Fix at #36107. In the mean time, I suggest using the workaround above.

kcreddy replied:
Thanks Dan!

efd6 commented on packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,180 @@
+                                                    "content_created_at": { "temp" : list_contents_resp_body}.collate("temp.contentCreated").max(),
+                                                    "next_page": (has(list_contents_resp.Header) && has(list_contents_resp.Header.NextPageUri) ) ? list_contents_resp.Header.NextPageUri[0] : (has(list_contents_resp.Header) && has(list_contents_resp.Header.Nextpageuri)) ? list_contents_resp.Header.Nextpageuri[0] : "",
+                                                    // keep fetching more if (nextpageuri exists) or (max time returned date != today's date)
+                                                    "want_more_content": ( (has(list_contents_resp.Header) && (has(list_contents_resp.Header.NextPageUri) && (list_contents_resp.Header.NextPageUri.size() > 0)) || (has(list_contents_resp.Header.Nextpageuri) && (list_contents_resp.Header.Nextpageuri.size() > 0)) ) || ( { "temp" : list_contents_resp_body}.collate("temp.contentCreated").max().split('T ...
```
```suggestion
                                                    "want_more_content": ( (has(list_contents_resp.Header) && (has(list_contents_resp.Header.NextPageUri) && (list_contents_resp.Header.NextPageUri.size() > 0)) || (has(list_contents_resp.Header.Nextpageuri) && (list_contents_resp.Header.Nextpageuri.size() > 0)) ) || ( { "temp" : list_contents_resp_body}.collate("temp.contentCreated").max().split('T')[0] != now.format(time_layout.RFC3339).split('T')[0] ) )
```

Recommend using a custom time format string without the zone if you don't need it, `s/time_layout.RFC3339/"2006-01-02"/g` (or when we are building on a later version of Go and mito is updated, `s/time_layout.RFC3339/time_layout.DateOnly/g`).

efd6 replied:
It just [calls `time.Time.Format`, and takes a string](https://github.com/elastic/mito/blob/3a86f09b8d5f3872229eb66129a863222d2924e6/lib/time.go#L201-L211). Those globals are just strings.

efd6 commented on packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,180 @@
+                                            has(list_contents_resp.StatusCode) 
+                                            && list_contents_resp.StatusCode == 200 
+                                            && (list_contents_resp.Request.URL.parse_url().RawQuery.parse_query().endTime[0].split('T')[0] != now().format(time_layout.RFC3339).split('T')[0]) 
+                                            ) ? true : false
```
```suggestion
                                            )
```

kcreddy replied:
Updated as suggested

efd6 commented on packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,180 @@
+                                state.cursor.content_types_state_as_list.filter(e, e.content_type == content_type)[0].content_created_at
+                            :
+                                (
+                                    (now() - duration("168h")).format(time_layout.RFC3339)
```
```suggestion
                                    (now - duration("168h")).format(time_layout.RFC3339)
```

kcreddy replied:
Updated to `now`

efd6 commented on packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,180 @@
+                (has(start_subs_resp_body.status) && start_subs_resp_body.status == 'enabled') || (has(start_subs_resp_body.error) && has(start_subs_resp_body.error.code) && start_subs_resp_body.error.code == "AF20024") ?
+                    // When start-subscription API returns success or if already started subscription,
+                    (request("GET", 
+                        ( has(state.want_more) && state.want_more && has(state.cursor) && has(state.cursor.content_types_state_as_list) && (state.cursor.content_types_state_as_list.filter(e, e.content_type == content_type)[0].next_page != "") ) ?
```
Since you define `state.want_more: false` in the initial state, you don't need the `has(state.want_more)` here.

kcreddy replied:
Makes sense 👍🏼 Updated as suggested

efd6 commented on packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,180 @@
+                            :
+                                // initial run when no cursor state exists i.e., polling from initial_interval
+                                state.base.list_contents_url + content_type + "&PublisherIdentifier={{azure_tenant_id}}"  
+                                    + "&startTime=" + (now() - duration(state.base.list_contents_start_time)).format(time_layout.RFC3339) 
```
```suggestion
                                    + "&startTime=" + (now - duration(state.base.list_contents_start_time)).format(time_layout.RFC3339) 
```

kcreddy replied:
Updated to `now`

efd6 commented on packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,180 @@
+                                // initial run when no cursor state exists i.e., polling from initial_interval
+                                state.base.list_contents_url + content_type + "&PublisherIdentifier={{azure_tenant_id}}"  
+                                    + "&startTime=" + (now() - duration(state.base.list_contents_start_time)).format(time_layout.RFC3339) 
+                                    + "&endTime=" + ((now() - duration(state.base.list_contents_start_time) + duration("24h")).as(calc_end_time, 
```
```suggestion
                                    + "&endTime=" + ((now - duration(state.base.list_contents_start_time) + duration("24h")).as(calc_end_time, 
```

kcreddy replied:
Updated to `now`

efd6 commented on packages/o365/data_stream/audit/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -24,10 +28,21 @@ processors:
+      on_failure:
+        - append:
+            field: error.message
+            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag fail-{{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
```
```suggestion
            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
```

kcreddy replied:
Updated as suggested

jamiehynds commented:
> Hey @jamiehynds, Below is how the upgrade scenario is going to look like for the user:
> 
>  O365-upgrade.1.mp4 
> Basically, the user will have to upgrade the integration policy too once the integration is upgraded to latest. Once that integration policy is updated, and since the new CEL input is enabled by default, the user will have to provide details for the CEL input to start. I have updated the title and description in the previous o365audit input to `Deprecated`. I also updated the `README` docs with the steps to follow when upgrading the integration from `< 1.17.0` to `>= 1.17.0`. Let me know if anything is missing or needs to be added/changed.
> 
> Also, during the tests, I noted that duplicates are not getting created. So, we are good there.
> 
> I didn't use the Graph API, since not all of the Office 365 and Azure AD activity events are available under Graph API. We are still using the [Office 365 Management Activity API](https://learn.microsoft.com/en-us/office/office-365-management-api/office-365-management-activity-api-reference) which is still actively supported.

@amitkanfer not sure if there's any improvements planned on the Fleet side, but this is a good example of a case where we have to make a signifiant change to an integration (changing an input) and difficult to inform users on the impact of the change. It's also a challenge to guide users on the steps to follow to ensure ingestion continues post-upgrade. In this case they have to disable the old data stream and re-enter their creds in a new data stream. While we try to keep these types of changes to a minimum, we find ourselves making them incases where a vendor deprecates an API or we have to switch inputs within packages. No action needed now, just an FYI.

kcreddy commented:
Hey @jamiehynds 
Update - the version is now `1.18.0`. So, the old o265audit input is now at `1.17.0`

> When a user upgrades the the integration the 'upgrade integration policies' option is selected, but in this case the policy isn't updated. It's probably a bit misleading, but not sure we can solve for that.
After upgrading the integration, a user has to go to integration policies to enable the new input and then upgrade the policy. I worry that users may not inattentively know to go to integration policies after the integration upgrade.

That is in fact correct and is misleading the users to think that upgrade is successful without actually upgrading the integration policy. But I am not sure what we can do here rather than providing a README documentation. Maybe the Fleet UI could redirect the user automatically to `integration policies` tab when there is such `Review field conflicts` warning comes up.

> I don't think people will know what CEL is. Could we adjust the wording of the new datastream to 'Collect Office 365 Audit Logs via Graph API' (or whatever API we're hitting)?

I updated the manifest.yml title to `Collect Office 365 audit logs via the Management Activity API using CEL Input`

> Am I correct in saying v1.17 of the integration will not include the deprecated input? So for new users using the O365 integration for the first time, will only have the option to use the CEL input, or will they also see the deprecated input too?

New version `v1.18` includes the deprecated input as well, but is disabled by default. So, new users also see the deprecated input as well. The CEL input is enabled by default. So, they need to configure the required parameters to save and make it run.

-----------

Also, I have now recorded a new video where I totally removed the deprecated input o365audit from the manifest.yml. Even now the users will have to manually upgrade the integration policy after upgrading the integration. But inside the policy UI, there is no deprecated input. It is bit cleaner, but the drawback could be that if something went wrong and users would like to continue reusing the old input, they would need to downgrade the integration version.

Here's the video of how it would look like without the deprecated input. Please let me know which of these you prefer, if the new integration version needs to contain or not contain the old deprecated input.

https://github.com/elastic/integrations/assets/11301409/fa1c4d35-de5f-4170-9b33-2d18d88191bb

efd6 commented on packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -138,7 +141,7 @@ program: |
                             :
                                 (
-                                    (now() - duration("168h")).format(time_layout.RFC3339)
+                                    (now - duration("168h")).format(time_layout.RFC3339)
```
Do we want to have the "168h" parameterisable?

kcreddy replied:
True. Updated it to value containing initial interval

jamiehynds commented:
> Also, I have now recorded a new video where I totally removed the deprecated input o365audit from the manifest.yml. Even now the users will have to manually upgrade the integration policy after upgrading the integration. But inside the policy UI, there is no deprecated input. It is bit cleaner, but the drawback could be that if something went wrong and users would like to continue reusing the old input, they would need to downgrade the integration version.
> 
> Here's the video of how it would look like without the deprecated input. Please let me know which of these you prefer, if the new integration version needs to contain or not contain the old deprecated input.

@kcreddy while removing the deprecated input is certainly cleaner, I think given the popularity of the integration and the extent the the httpjson to CEL change, it makes more sense to have an escape hatch available, at least for a release or two as users transition to the upgraded integration. So I'd vote to keeping the deprecated input as fallback for now. WDYT @andrewkroh?

kcreddy commented:
> If the deprecated input remains for a release or two, then reverting to httpjson would be a matter or enabling httpjson and disabling cel via the existing toggle. The user would not need to enter any new config because the old config would be saved. Correct?

The old input for this package is `o365audit`. But yes, I verified that the old config is being saved. Hence the user can just disabled `CEL` input and enable the old `o365audit` input.

> If the httpjson input was removed, what's the level of difficulty in downgrading the package version if something goes wrong?

That would mean they would need to uninstall the integration and reinstall the older version. I found this tracking issue: https://github.com/elastic/kibana/issues/131801. Currently there does not seem to be an easier way to downgrade the package versions in Fleet.

andrewkroh commented:
Because downgrading is complicated and the user would have lost their existing configuration for the o365audit input, I think keeping the input around for a brief transition period will help if there is some unknown issue.

efd6 commented on packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,183 @@
+                                                    "content_created_at": { "temp" : list_contents_resp_body}.collate("temp.contentCreated").max(),
+                                                    "next_page": (has(list_contents_resp.Header) && has(list_contents_resp.Header.NextPageUri) ) ? list_contents_resp.Header.NextPageUri[0] : (has(list_contents_resp.Header) && has(list_contents_resp.Header.Nextpageuri)) ? list_contents_resp.Header.Nextpageuri[0] : "",
+                                                    // keep fetching more if (nextpageuri exists) or (max time returned date != today's date)
+                                                    "want_more_content": ( (has(list_contents_resp.Header) && (has(list_contents_resp.Header.NextPageUri) && (list_contents_resp.Header.NextPageUri.size() > 0)) || (has(list_contents_resp.Header.Nextpageuri) && (list_contents_resp.Header.Nextpageuri.size() > 0)) ) || ( { "temp" : list_contents_resp_body}.collate("temp.contentCreated").max().split('T ...
```
This can be made less expensive. The important thing here is that `||` and `&&` are not short circuiting, so the collation is always evaluated, even when the first or second operands of the first two disjunctions are true. Normally, I don't worry about this, but in this case, `collate` is reasonably expensive and so we should avoid it if possible.

There's notes about this [here](https://github.com/google/cel-spec/blob/master/doc/langdef.md#logical-operators).

kcreddy replied:
Thanks Dan for the suggestions. I will definitely take it as an enhancement.

efd6 commented on packages/o365/data_stream/audit/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,183 @@
+                                        ,
+                                        "next_page": "",
+                                        "want_more_content": (
+                                            has(list_contents_resp.StatusCode) 
+                                            && list_contents_resp.StatusCode == 200 
+                                            && (list_contents_resp.Request.URL.parse_url().RawQuery.parse_query().endTime[0].split('T')[0] != now().format("2006-01-02")) 
```
We can do a similar thing here.

efd6 reviewed (approved):
Minor issues only.

The CEL code issues can be a future optimisation is you prefer.

elasticmachine commented:
Package o365 - 1.18.0 containing this change is available at https://epr.elastic.co/search?package=o365
